sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala

```diff
@@ -257,16 +261,33 @@ class FileStreamSource(
    * equal to `end` and will only request offsets greater than `end` in the future.
    */
   override def commit(end: Offset): Unit = {
-    // No-op for now; FileStreamSource currently garbage-collects files based on timestamp
-    // and the value of the maxFileAge parameter.
+    val logOffset = FileStreamSourceOffset(end).logOffset
+
+    if (sourceOptions.cleanSource != CleanSourceMode.NO_OP) {
+      val files = metadataLog.get(Some(logOffset), Some(logOffset)).flatMap(_._2)
+      val validFileEntities = files.filter(_.batchId == logOffset)
+      logDebug(s"completed file entries: ${validFileEntities.mkString(",")}")
+      sourceOptions.cleanSource match {
+        case CleanSourceMode.ARCHIVE =>
+          validFileEntities.foreach(sourceCleaner.archive)
+
+        case CleanSourceMode.DELETE =>
+          validFileEntities.foreach(sourceCleaner.delete)
+
+        case _ =>
+      }
+    } else {
+      // No-op for now; FileStreamSource currently garbage-collects files based on timestamp
+      // and the value of the maxFileAge parameter.
+    }
+
```

Comment:
Maybe we can also purge metadata here to prevent growing infinitely given that Spark guarantees it will not happen to request offset equal or less than `logOffset` (I guess it would also apply to `getBatch`).